esx_service/vmdk_ops.py

```diff
@@ -844,24 +848,48 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     except ValidationError as ex:
         return err(str(ex))
 
-    if not datastore:
-        datastore = default_datastore
-
     if datastore and not vmdk_utils.validate_datastore(datastore):
         return err("Invalid datastore '%s'.\n" \
                    "Known datastores: %s.\n" \
                    "Default datastore: %s" \
                    % (datastore, ", ".join(get_datastore_names_list()), default_datastore))
 
-    # datastore is present and validated.
-    datastore_url = vmdk_utils.get_datastore_url(datastore)
+    if not datastore:
+        datastore_url = default_datastore_url
+        datastore = default_datastore
+        use_default_ds = True
+    else:
+        datastore_url = vmdk_utils.get_datastore_url(datastore)
+        use_default_ds = False
 
-    logging.debug("executeRequest: vm_uuid=%s, vm_name=%s, tenant_name=%s, tenant_uuid=%s, default_datastore_url=%s",
-                  vm_uuid, vm_name, tenant_uuid, tenant_name, default_datastore_url)
+    logging.debug("executeRequest: vm_uuid=%s, vm_name=%s, tenant_name=%s, tenant_uuid=%s, "
+                  "default_datastore_url=%s datastore_url=%s",
+                  vm_uuid, vm_name, tenant_uuid, tenant_name, default_datastore_url, datastore_url)
 
-    error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, datastore_url, cmd, opts)
-    if error_info:
-        return err(error_info)
+    if use_default_ds:
+        # first check whether it has privilege to default_datastore
+        error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, datastore_url, cmd, opts)
+        if error_info:
+            return err(error_info)
+    else:
+        # user passsed in volume with format vol@datastore
+        # check the privilege to that datastore
+        error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, datastore_url, cmd, opts)
+        # no privilege exist for the given datastore
+        # if the given datastore is the same as vm_datastore
+        # then we can check privilege against "__VM_DS"
+        if error_info == error_code_to_message[ErrorCode.PRIVILEGE_NO_PRIVILEGE] and datastore == vm_datastore:
+            error_info, tenant_uuid, tenant_name = auth.authorize(vm_uuid, auth_data_const.VM_DS_URL, cmd, opts)
+        if error_info:
+            return err(error_info)
+
+    # get_vol_path() need to pass in a real datastore name
+    if datastore == auth_data_const.VM_DS:
+        datastore = vm_datastore
+        # set datastore_url to a real datastore_url
+        # createVMDK() and removeVMDK() need to pass in
+        # a real datastore_url instead of url of __VM_DS
+        datastore_url = vm_datastore_url
```

Comment:
I put the logic for authorize check into one function. For line 886-892, I don't think it fits to the authorize_check function, and I will keep it in the executeRequest().